rust/wal3/README.md

```diff
@@ -0,0 +1,468 @@
+    - setsum:  The setsum of the log fragment.
+
+Invariants of the manifest:
+
+- The setsum of all fragments in a manifest plus `collected` must add up to the `setsum` of the
+  manifest.
+- fragments.seq_no is sequential.
+- fragment.start < fragment.limit for all fragments.
+- fragment.start is strictly increasing.
+- The range (fragment.start, fragment.limit) is disjoint for all fragments in a manifest.  No other
+  fragment will have overlap with log position.
+
+## Object Store Layout
+
+wal3 is designed to maximize object store performance of object stores like S3 because it writes
+logs in a way that scales.  Concretely, we leverage the behavior that S3 and similar services
+institute rate limiting per prefix.  For example, given the following log files in an S3 bucket,
+we will group fragments in groups of 5000 and the manifest will be in a separate prefix.
+
+The following shows numbers every 5000.  I'd zero-pad to 16 hex digits for the sequence number and
+bucket fragments in groups of 4096 so the bits align and look pretty in the seqno prefix.
+
+```text
+wal3/SeqNo=    0/0000000000000000.parquet
+wal3/SeqNo=    0/0000000000000001.parquet
+wal3/SeqNo=    0/0000000000000002.parquet
+...
+wal3/SeqNo=    0/0000000000004999.parquet
+wal3/SeqNo= 5000/0000000000005000.parquet
+...
+wal3/SeqNo=10000/0000000000010000.parquet
+...
+wal3/SeqNo=15000/0000000000015000.parquet
+...
+wal3/Manifest/MANIFEST.json
+```
+
+## Writer Arch Diagram
+
+```text
+┌─ Writer ──────────────────────────────────────────────────┐
+│  ┌─────────────────────────────┐  ┌────────────────────┐  │
+│  │ fragment                    │  │ manifest           │  │
+│  │ manager                     │  │ manager            │  │
+│  │                             │  │                    │  │
+│  │ - new                       │  │ - new              │  │
+│  │ - push_work                 │  │ - assign_timestamp │  │
+│  │ - take_work                 │  │ - apply_delta      │  │
+│  │ - wait_for_writable         │  │                    │  │
+│  │ - update_average_batch_size │  │                    │  │
+│  │ - finish_write              │  │                    │  │
+│  └─────────────────────────────┘  └────────────────────┘  │
+└───────────────────────────────────────────────────────────┘
+```
+
+The write path is:
+
+2.  The writer calls `push_work` to submit work to the fragment manager.  This enqueues the work.
+3.  The writer calls `take_work` from the fragment manager.  If there is a batch of sufficient size
+    and a free fragment, it will assign the work to that fragment and return the work to be written.
+    Go to 4.  If there is no batch, skip to step 3a.
+    a.  Enqueue and wait for some other task to signal that the work is ready.  Go to 6.
+4.  Flush the work from take_work to object storage.  This will call assign-timestamp on the
+    manifest manager.
+5.  The writer creates a delta to the manifest---the new fragment and its setsum---and calls
```

Comment:
Are manifests partitioned in any way? Or are all writers submitting deltas to the single manifest manager?

Reply:
Each chroma collection is 1:1 with a manifest.

Reply:
I guess this isn't obvious in the description of the Manifest above. Should we make this more clear as another "invariant" of the manifest? The reason I asked is because I was reading it as multiple writers all contending for operations on a single manifest.